### PR TITLE
allow autonat v2 to dial all ips

### LINKED ISSUE #536: Proposal: allow AutoNAT to dial all IP addresses, without risking amplification attacks
Our AutoNAT spec currently says:
> In order to prevent attacks like the one described in [RFC 3489, Section 12.1.1](https://www.rfc-editor.org/rfc/rfc3489#section-12.1.1) (see excerpt below), implementations MUST NOT dial any multiaddress unless it is based on the IP address the requesting node is observed as. This restriction as well implies that implementations MUST NOT accept dial requests via relayed connections as one can not validate the IP address of the requesting node.
>
>> RFC 3489 12.1.1 Attack I: DDOS Against a Target
>>
>> In this case, the attacker provides a large number of clients with the same faked MAPPED-ADDRESS that points to the intended target. This will trick all the STUN clients into thinking that their addresses are equal to that of the target. The clients then hand out that address in order to receive traffic on it (for example, in SIP or H.323 messages). However, all of that traffic becomes focused at the intended target. The attack can provide substantial amplification, especially when used with clients that are using STUN to enable multimedia applications.

The intention is to prevent an amplification attack against a target, and was designed with the properties of the STUN protocol in mind.

This is not the only way to prevent amplification attacks though. We can also just make the attack expensive enough such that it becomes unattractive. For example, we could require the requester to send a non-trivial amount of data, on the order of 10-100 kB. This would make AutoNAT completely uninteresting for an amplification attack, since a libp2p handshake is a lot smaller than 10 kB, and thus no amplification can be achieved.

On the other hand, for the AutoNAT server, receiving 10-100 kB is cheap enough to process that we're not placing a too large burden on the server. To further lighten the load on the server, this could be designed as a "retry" mechanism (borrowing QUIC terminology here): Only if the IP address doesn't match would the server request the client to send this additional data.

Obviously, as this is a new design, this would need to be incorporated into AutoNAT v2 (#503).

@sukunrt Thoughts?

### PR DESCRIPTION
**Pull request by sukunrt**
Closes #536

### CODE REVIEW

File: autonat/autonat-v2.md [hunk 1]
```diff
@@ -95,25 +104,23 @@ the message. This includes inability to decode the requested address to dial.
...
-> In this case, the attacker provides a large number of clients with the same
-> faked MAPPED-ADDRESS that points to the intended target. This will trick all
-> the STUN clients into thinking that their addresses are equal to that of the
-> target. The clients then hand out that address in order to receive traffic on
-> it (for example, in SIP or H.323 messages). However, all of that traffic
-> becomes focused at the intended target. The attack can provide substantial
-> amplification, especially when used with clients that are using STUN to enable
-> multimedia applications.
+Upon receiving a `DialDataRequest` the client starts transferring `numBytes`
+number of bytes to the server.
```
Comment (sukunrt) [inline autonat/autonat-v2.md:121]: Should I add a step where the client first sends a message indicating that it has accepted/rejected the cost? 
@marten-seemann

Comment (sukunrt) [reply autonat/autonat-v2.md]: i added this step.

File: autonat/autonat-v2.md [hunk 2]
```diff
@@ -36,29 +36,38 @@ Implementations can generate low priority address guesses and add them to
...
 
 Compared to `autonat v1` there are two major differences
-1. `autonat v1` allowed testing reachability for the node. `autonat v2` allows
-testing reachability for an individual address
+1. `autonat v2` allows testing reachability for an individual address. `autonat
+   v1` allowed testing reachability for the node. 
 2. `autonat v2` provides a mechanism for nodes to verify whether the peer
 actually successfully dialled an address.
+3. `autonat v2` provides a mechanism for nodes to dial an ip address different
+from the requesting nodes observed ip address without risking amplification
```
Comment (marten-seemann) [inline autonat/autonat-v2.md:44]:
```suggestion
from the requesting node's observed ip address without risking amplification
```

File: autonat/autonat-v2.md [hunk 3]
```diff
@@ -167,7 +180,15 @@ message DialRequest {
...
-    fixed64 nonce = 1;
+  fixed64 nonce = 1;
+}
+
+message DialDataRequest {
+  uint64 numBytes = 1;
+}
+
+message DialDataResponse {
+  bool accepted = 1;
```
Comment (marten-seemann) [inline autonat/autonat-v2.md:191]: Why not put the bytes into this protobuf? Then we wouldn't need to break the framing to send these bytes.

Comment (sukunrt) [reply autonat/autonat-v2.md:191]: That would require memory of the order of 100kb for the protobufs on both ends, right? 
The current version can be implemented using a smaller buffer.

I don't have good intuition here, do you think this is fine?

Comment (sukunrt) [reply autonat/autonat-v2.md:191]: @marten-seemann friendly ping

Comment (MarcoPolo) [reply autonat/autonat-v2.md:191]: The server is going to discard these bytes right away anyways. No point to allocate the space to put the bytes in the message. I agree with the initial design. (Sorry for the drive-by comment, I'll defer to whatever you all decide).

File: autonat/autonat-v2.md [hunk 4]
```diff
@@ -95,25 +104,26 @@ the message. This includes inability to decode the requested address to dial.
...
 
-In order to prevent attacks like the one described in [RFC 3489, Section
-12.1.1](https://www.rfc-editor.org/rfc/rfc3489#section-12.1.1) (see excerpt
-below), implementations MUST NOT dial any multiaddress unless it is based on the
-IP address the requesting node is observed as. This restriction as well implies
-that implementations MUST NOT accept dial requests via relayed connections as
-one can not validate the IP address of the requesting node.
+Upon receiving a `DialDataRequest` message the client responds with
+`DialDataResponse` message with a boolean `accepted` indicating if it accepted
+the request. If `accepted` is `false`, the `DialRequest` is considered complete
```
[the remark this answers is not in the document]
Comment (sukunrt) [reply autonat/autonat-v2.md]: I am accepting this. This looks simpler to me. Even if we add a different cost for dial in future, say some POW like mechanism, the client can just reject the charge by resetting the stream.

File: autonat/autonat-v2.md [hunk 5]
```diff
@@ -107,26 +120,24 @@ peer that prevented it from completing the request.
...
-> In this case, the attacker provides a large number of clients with the same
-> faked MAPPED-ADDRESS that points to the intended target. This will trick all
-> the STUN clients into thinking that their addresses are equal to that of the
-> target. The clients then hand out that address in order to receive traffic on
-> it (for example, in SIP or H.323 messages). However, all of that traffic
-> becomes focused at the intended target. The attack can provide substantial
-> amplification, especially when used with clients that are using STUN to enable
-> multimedia applications.
+On receiving a `DialRequest`, the server selects the first address it is capable
+of dialing. If this selected address has a ip different from the clients
```
Comment (marten-seemann) [inline autonat/autonat-v2.md:132]:
```suggestion
of dialing. If this selected address has a ip different from the client's
```

File: autonat/autonat-v2.md [hunk 6]
```diff
@@ -107,26 +120,24 @@ peer that prevented it from completing the request.
...
+observed ip, the server sends a `DialDataRequest` message with `numBytes` set to
+a sufficiently large value on the `/libp2p/autonat/2.0.0/dial-request` stream
 
+Upon receiving a `DialDataRequest` message, the client decides whether to accept
+or reject the cost of dial. If the client rejects the cost, the client resets
+the stream and the `DialRequest` is considered complete. If the client accepts
```
Comment (marten-seemann) [inline autonat/autonat-v2.md:138]: Complete? Not aborted?

**Review by marten-seemann — APPROVED** [review]
LGTM.

Nit: I think it's spelled IP and not ip.

File: autonat/autonat-v2.md (hunk 6 above)
Comment (sukunrt) [reply autonat/autonat-v2.md]: that's a much better term

**Comment by sukunrt** [comment]
>Nit: I think it's spelled IP and not ip.

oops, thanks!

**Pull request merged**